File: nodejs-tools/nodejs-compiler-tools/formatter-xml/src/format_xml_program.ts
import { XmlProgram } from '@vighnesh153/parser-xml';
import { formatXmlExpression } from './format_xml_expression';

type FormatXmlProgramConfig = {
  readonly program: XmlProgram;
  readonly indentation: number;
  readonly sortAttributes: boolean;
};

export function formatXmlProgram({ program, indentation, sortAttributes }: FormatXmlProgramConfig): string {
  const stringBuilder: string[] = [];

  for (const statement of program.statements) {
    stringBuilder.push(
      formatXmlExpression({ expression: statement, indentation, indentationLevel: 0, sortAttributes })
    );
  }

  return (
    stringBuilder.join('\n') +
    // empty line at the end
    '\n'
  );
}

File: jvm-tools/XmlFormatter/src/main/kotlin/dev/vighnesh153/xml/formatter/fmt/FormatXmlProgram.kt
package dev.vighnesh153.xml.formatter.fmt

import dev.vighnesh153.xml.formatter.ast.XmlProgram

internal fun XmlProgram.format(
    indentation: Int,
    sortAttributes: Boolean,
): String = statements.joinToString("\n") {
    formatXmlExpression(
        expr = it,
        indentationLevel = 0,
        indentation = indentation,
        sortAttributes = sortAttributes
    )
} + "\n"
